ethier/sign.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/divergencetech/ethier/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/spf13/cobra"
+)
+
+var signer *eth.Signer
+
+func init() {
+	var signCmd = &cobra.Command{
+		Use:   "sign",
+		Short: "Signs messages from stdin using an ECDSA signer.",
+	}
+
+	rootCmd.AddCommand(signCmd)
+	signCmd.PersistentFlags().String("mnemonic", "", "Path to file containing the mnemonic, password and account number (newline separated) for key derivation")
+
+	var signAddrCmd = &cobra.Command{
+		Use:     "addresses",
+		Short:   "Signs addresses from stdin using an ECDSA signer.",
+		PreRunE: getSigner,
+		RunE:    signAddresses,
+	}
+
+	signCmd.AddCommand(signAddrCmd)
+}
+
+type SignedAddress struct {
+	Address   string `json:"address"`
+	Signature string `json:"signature"`
+}
+
+func getSignerFromMnemonic(filepath string) (*eth.Signer, error) {
+	dat, err := os.ReadFile(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("Read mnemonic: %v", err)
+	}
+
+	lines := strings.Split(string(dat), "\n")
+
+	mnemonic := lines[0]
+	var password string = ""
+	var account uint = 0
+
+	if len(lines) > 1 {
+		password = lines[1]
+	}
+
+	if len(lines) > 2 {
+		if lines[2] != "" {
+			acc, err := strconv.Atoi(lines[2])
+			if err != nil {
+				return nil, fmt.Errorf("Converting derivation account: %v", err)
+			}
+			account = uint(acc)
+		}
+	}
+
+	signer, err := eth.DefaultHDPathPrefix.SignerFromSeedPhrase(
+		mnemonic, password, account,
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("Generate signer: %v", err)
+	}
+	return signer, nil
+}
+
+func getSigner(cmd *cobra.Command, args []string) (retErr error) {
+	mnemonicPath, err := cmd.Flags().GetString("mnemonic")
+	if err != nil {
+		log.Fatalf("Get mnemonics flag: %v", err)
+	}
+
+	if mnemonicPath != "" {
+		signer, err = getSignerFromMnemonic(mnemonicPath)
+		if err != nil {
+			log.Fatalf("Generate signer from mnemonic: %v", err)
+		}
+	} else {
+		signer, err = eth.NewSigner(256)
+		if err != nil {
+			log.Fatalf("Generate new signer: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// sign addresses generates a new signer or derives it from a given mnemonic
+// to sign a list of addresses
+func signAddresses(_ *cobra.Command, args []string) (retErr error) {
+	defer func() {
+		if retErr != nil {
+			retErr = fmt.Errorf("signing: %w", retErr)
+		}
+	}()
```

Comment:
This pattern isn't necessary here, and is actually not terribly common in Go. I used it in `ethier gen` to tag the errors with additional metadata about the file being compiled, but that's redundant here. You can also remove the `retErr` from the function signature as it's only needed with the `defer`.

I saw you'd put a comment in an earlier version about `%w`: https://pkg.go.dev/errors. I always use this when tagging an error in a deferred function so the underlying error is available in its original form. I don't suggest using it all the time because that makes internal errors part of the public API of the code you're writing (e.g. if you get an error from a different package and propagate it with `%w` then you make their public API part of yours, so `%v` is a better default).

Reply:
I see. Thanks for the elaboration.